src/mds/Server.cc

```diff
@@ -3899,6 +3899,11 @@ void Server::handle_client_lookup_ino(MDRequestRef& mdr,
     return _lookup_snap_ino(mdr);
 
   inodeno_t ino = req->get_filepath().get_ino();
+  auto _ino = ino.val;
+  if (MDS_INO_IS_STRAY(_ino) || MDS_INO_IS_MDSDIR(_ino)) {
```

Comment:
How about this instead?
```
/* Don't allow client to look up inodes that are for MDS internal usage */
if (_ino < MDS_INO_SYSTEM_BASE && _ino != MDS_INO_ROOT) {
``` 
...that should catch any inode numbers in the MDS "private" range. Of course, if there is a legit reason to allow lookups of some of these inodes, then we will need to carve some others out.

Reply:
That also looks good. Let's see first what @ukernel says about  https://github.com/ceph/ceph/pull/40389#issuecomment-806632075

Reply:
@ukernel: ping? Do you have any thoughts on the above? I'd like to add similar checks to the kernel (along with some stack traces so we can see what codepaths lead to these calls).

Reply:
There has another issue [1], which is the mdsdir `0x101` had issued caps to 3 clients and it prevented expiring the mdsdir inode in `mds.0`. 

In this case they were using both kclient and userspace client at the same time, so we also need a similar checks in `client/Client.cc`.

[1] https://tracker.ceph.com/issues/50112

Reply:
I think this looks good, but I'd prefer we use the condition I suggested above. That's what we're doing in the client, so we probably ought to do the same in the MDS.

Reply:
Done.